File: Problems/No.680_Valid_Palindrome_II/valid_palindrome_II.py
class Solution(object):
    def validPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """

        # palindrome chance of deleting one character
        def PalindromeChance(L, R):
            while (L < R):
                if (s[L] == s[R]):
                    L += 1
                    R -= 1
                else:
                    return False
            return True
        

        # parameters
        pLeft, pRight = 0, len(s)-1
    
        # start computing
        while (pLeft < pRight):
            if (s[pLeft] == s[pRight]):
                pLeft += 1
                pRight -= 1
            else:
                return (PalindromeChance(pLeft+1, pRight) or PalindromeChance(pLeft, pRight-1))
        
        return True

File: Problems/No.680_Valid_Palindrome_II/valid_palindrome_II.cpp
class Solution {
public:

    // palindrome chance of deleting one character
    bool PalindromeChance(std::string s, int L, int R) {

        while (L < R) {

            if (s[L] == s[R]) {

                L++;
                R--;

            }
            else return false;

        }

        return true;

    }

    bool validPalindrome(string s) {
        
        // parameters
        int pLeft = 0, pRight = s.length()-1;

        // start computing
        while (pLeft < pRight) {

            if (s[pLeft] == s[pRight]) {

                pLeft++;
                pRight--;

            }
            else return (PalindromeChance(s, pLeft+1, pRight) || PalindromeChance(s, pLeft, pRight-1));
        }

        return true;

    }
};
